assets/pages/new-contributions-landing/components/ContributionForm.jsx

```diff
@@ -0,0 +1,202 @@
+  onError: (error) => { dispatch(paymentFailure(error)); },
+});
+
+// ----- Functions ----- //
+
+const getAmount = (formElements: Object) =>
+  parseFloat(formElements.contributionAmount.value === 'other'
+    ? formElements.contributionOther.value
+    : formElements.contributionAmount.value);
+
+// ----- Event handlers ----- //
+
+let stripeReady = null;
+
+const onSubmit = (e) => {
+  e.preventDefault();
+
+  const { elements } = e.target;
+  const amount = getAmount(elements);
+  const email = elements.namedItem('contributionEmail').value;
+
+  if (stripeReady) {
+    stripeReady.then(() => {
+      openDialogBox(amount, email);
+    });
+  }
+};
+
+
+// ----- Render ----- //
+
+function setupStripe(formElement, props) {
+  const {
+    abParticipations,
+    currency,
+    referrerAcquisitionData,
+    optimizeExperiments,
+    thankYouRoute,
+  } = props;
+
+  const callback = createTokenCallback({
+    abParticipations,
+    currencyId: currency,
+    referrerAcquisitionData,
+    optimizeExperiments,
+    getData: () => {
+      const elements: any = formElement.elements; // eslint-disable-line prefer-destructuring
+      const firstName = elements.contributionFirstName.value;
+      const lastName = elements.contributionLastName.value;
+      const email = elements.contributionEmail.value;
+      const amount = getAmount(elements);
+
+      return {
+        user: {
+          fullName: `${firstName} ${lastName}`,
+          email,
+        },
+        amount,
+      };
+    },
+    onSuccess: () => {
+      trackConversion(abParticipations, thankYouRoute);
+      props.onSuccess();
+    },
+    onError: props.onError,
+  });
+
+  stripeReady = setupStripeCheckout(callback, null, currency, props.isTestUser);
+}
+
+function ContributionForm(props: PropTypes) {
+  const {
+    countryGroupId,
+    selectedCountryGroupDetails,
+    currency,
+    thankYouRoute,
+  } = props;
+
+  return props.done ?
+    <Redirect to={thankYouRoute} />
+    : (
+      <div className="gu-content__content">
+        <h1>{countryGroupSpecificDetails[countryGroupId].headerCopy}</h1>
+        <p className="blurb">{countryGroupSpecificDetails[countryGroupId].contributeCopy}</p>
+        <ErrorMessage message={props.error} />
+        <form
+          ref={(el) => {
+          if (el) {
+            setupStripe(el, props);
+          }
+        }}
+          className={classNameWithModifiers('form', ['contribution'])}
+          onSubmit={onSubmit}
+        >
+          <NewContributionType />
+          <NewContributionAmount
+            countryGroupId={countryGroupId}
+            countryGroupDetails={selectedCountryGroupDetails}
+            currency={currency}
+          />
+          <NewContributionTextInput
+            id="contributionFirstName"
+            name="contribution-fname"
+            label="First Name"
+            value={window.guardian.user && window.guardian.user.firstName ? window.guardian.user.firstName : null}
```

Comment:
won't this make the input's uneditable? as the value of `window.guardian.user.firstName` will never change

Reply:
I've just tried this locally and it isn't the case, and it's because there isn't anything on the form at the moment that changes the state yet, which would cause a re-render and write over anything written in the boxes. This is why we have to use the state to set the value of inputs

Reply:
Nope. Thanks to the DOM diffing, these elements will never be re-rendered. However I agree that in certain circumstances you'd be forced to keep track of the field values 😭

Reply:
What are these circumstances? If we don't need to then that's great! I'll have to get you to explain  this DOM-diffing to me tomorrow

Reply:
In short, yes. I just think it is a terrible design on the part of React/Redux to impose that kind of constraint, but it doesn't mean I will bend to it 😉 There is no reason to duplicate a perfectly working state machine (the form) on the client side, that's why these properties are immutable (there is alas no such thing in JS except through discipline). Duplication leads to complecting (having to deal with `onchange` and `onblur`) and bugs like the forever hanging page that happened the other day.

Reply:
> What are these circumstances?

When the rendering of the element is conditional, you can either abstain from supplying a value or you are forced to track the value in the state. So in this case:
```
{condition ? <input value={intialValue} /> : null }
```
That's fair enough, as there's an easy workaround:
```
<input value={initialValue} hidden={!condition} />
```